### ISSUE TITLE
MSAL isn't able to decrypt credentials (UWP)

### ISSUE THREAD

**Issue by tipa** [AzureAD/microsoft-authentication-library-for-dotnet#889]
**Which Version of MSAL are you using ?**
MSAL 2.7.0

**Platform**
UWP

**Is this a new or existing app?**
a. The app is in production, and I have upgraded to a new version of MSAL

```csharp
 var accounts = await clientApplication.GetAccountsAsync();
```

**Expected behavior**
No exception

**Actual behavior**
An exception is thrown - rarely! I am unable to provide a project to easily reproduce this issue, but it has happened recurrently for some of my users now.

`System.Exception: The specified data could not be decrypted. (Exception from HRESULT: 0x8009002C)   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x21   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x5c   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task) + 0x44   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task) + 0x1c   at Microsoft.Identity.Client.Platforms.uap.UapCryptographyManager.<>c__DisplayClass11_0`1.<<RunAsyncTaskAndWait>b__0>d.MoveNext() + 0xed   at Microsoft.Identity.Client.Platforms.uap.UapCryptographyManager.RunAsyncTaskAndWait[T](Task`1) + 0x128   at Microsoft.Identity.Client.Platforms.uap.UapCryptographyManager.Decrypt(Byte[]) + 0x86   at Microsoft.Identity.Client.Platforms.uap.UapTokenCacheAccessor.GetCacheValue(ApplicationDataCompositeValue) + 0x1f0   at Microsoft.Identity.Client.Platforms.uap.UapTokenCacheAccessor.GetAllRefreshTokensAsString() + 0xef   at Microsoft.Identity.Client.TokenCache.GetAllRefreshTokensForClient(RequestContext) + 0x90   at Microsoft.Identity.Client.TokenCache.GetAccounts(String, Boolean, RequestContext)`

**Comment by jmprieur** [comment]
@tipa: does it only happen with MSAL 2.7 ?

**Comment by tipa** [comment]
No, it also happened with 2.6.2 and older versions (the first log that I have is from April 28th '18, at that time v1.1.2-preview0008 was the latest MSAL version.

**Comment by bgavrilMS** [comment]
MSAL 3 changes the way we persist the token cache on UWP. We now use DPAPI to encrypt and then we save the file to LocalData. We also give developers the possiblity of overriding this behaviour by implemeting their own token cache serialization.

Could you try with MSAL 3 please (preview is out).

**Comment by tipa** [comment]
Does this mean my existing users have to go through the whole login process again?

**Comment by bgavrilMS** [comment]
No, your app should be able to preserve SSO. We actually store the token cache in 2 places, one that is MSAL specific and one that is ADAL specific, because we expect people to be able to gradually migrate from ADAL to MSAL. 

Anyway, a nice side effect from this is that we can be a bit liberal in making changes to where we store the MSAL cache, because the ADAL cache will ensure SSO.

**Comment by tipa** [comment]
Great - thanks! I will use MSAL 3 in my apps then and update this issue if the problem shows up again.

**Comment by jmprieur** [comment]
@tipa : you might want to read https://aka.ms/msal-net-3x

**Comment by jennyf19** [comment]
Fixed in [MSAL 3.0.0-preview release](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/releases/tag/3.0.0-preview)
